### ISSUE TITLE
"context deadline exceeded" when requesting pachctl version

### ISSUE THREAD

**Issue by thedrow** [pachyderm/pachyderm#1881]
I recently upgraded my dev machine to use the most recent pachctl. It seems like something is wrong when I request the version.
```
pachctl version
COMPONENT           VERSION             
pachctl             1.4.6               
context deadline exceeded
```

**Comment by thedrow** [comment]
I tried using the `--no-metrics` option as well but it still seems like some other I/O happens.

**Comment by JoeyZwicker** [comment]
Hi Omer, I know this is a bit of a silly question, but is your pachd pod actually up and running in this case and have you forwarded the port? I haven't seen this response except when there's literally no pachd connection.

What's `kubectl get all` show?

**Comment by thedrow** [comment]
No it wasn't. There shouldn't be an error in any case.

**Comment by thedrow** [comment]
It seems that "context deadline exceeded" is printed whenever pachctl is not able to connect to the server when invoking any command that communicates with it. It's not a very clear error.

**Comment by jdoliner** [comment]
@thedrow agreed that this is not a very clear error, we should catch it and output something clearer.

**Comment by thedrow** [comment]
This should be fixed in 1.5 right?

**Comment by JoeyZwicker** [comment]
hi @paologf `minikube delete && minikube start` is a great way to reset everything. You'll need to do `pachctl deploy local` after that before `port-forward` will work. You can also upgrade to v1.5.0 by following our installation instructions. 

'context deadline exceeded` usually comes from a command being unable to connect to the pachd node -- generally because either your cluster isn't up at all or pachd has crashed and is restarting. You can use `kubectl get all`to see a list of pods that are running and their states. What does that show in your cluster?

**Comment by paologf** [comment]
Hi @JoeyZwicker, 
thanks for the reply.

I was redoing the steps before reply you and this time it works.
I'm thinking that was only a matter of timing between `pachctl deploy local ` and `port-forward `.
Indeed the `kubectl get all` message changed from `READY 0/1` to `READY 1/1`.

Thank you.

**Comment by JoeyZwicker** [comment]
@paologf Yeah that makes sense. minikube has to pull the pachd image from dockerhub when you deploy so it takes about 30-60 seconds depending on your internet connection. Glad your cluster is working now and don't hesitate to post more questions if you run into anything.

**Comment by JoeyZwicker** [comment]
@jdoliner re: your above comment, can we catch this and print something along the lines of "cannot connect to pachd. Please check that pachd pod is running with `kubectl get all`"?

**Comment by merl-dev** [comment]
same problem:

```
$ kubectl get all
NAME                        READY     STATUS    RESTARTS   AGE
po/etcd-281005231-jrpxr     1/1       Running   0          37m
po/pachd-1390770499-z44c9   1/1       Running   3          37m

NAME             CLUSTER-IP      EXTERNAL-IP   PORT(S)                       AGE
svc/etcd         100.70.194.1    <nodes>       2379:30638/TCP                37m
svc/kubernetes   100.64.0.1      <none>        443/TCP                       1h
svc/pachd        100.64.20.148   <nodes>       650:30650/TCP,651:30651/TCP   37m

NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/etcd    1         1         1            1           37m
deploy/pachd   1         1         1            1           37m

NAME                  DESIRED   CURRENT   READY     AGE
rs/etcd-281005231     1         1         1         37m
rs/pachd-1390770499   1         1         1         37m

```

but when I run:
```
$ pachctl version
COMPONENT           VERSION             
pachctl             1.5.0               
context deadline exceeded
```

This wasn't happening earlier (about 2-3 minutes after firing up the cluster)

**Comment by JoeyZwicker** [comment]
hi @merl-dev, sorry you're hitting this as we've generally been unable to reproduce. 

The only times we've been able to see this error is when you can't connect to pachd. 

You said it wasn't happening 2-3 minutes after firing up the cluster. By that, do you mean that `pachctl version` returned correctly? 

The pod clearly looks like it's running -- are the restart numbers going up, indicating a crashloop backoff? And you've done port-forwarding (and the pf process didn't die), right? 

Is this is minikube or on a cloud provider?

If none of the above issues seem to oviously be the culprit, it may be faster to help figure out what's going on live so feel free to join our [public channel](slack.pachyderm.io).

**Comment by sjezewski** [comment]
stale

**Comment by JoeyZwicker** [comment]
We still need to make this error message clearer. Context deadline exceeded doesn't mean anything and users hit this a lot.

**Comment by radu-adrian-moldovan** [comment]
Hi guys, run into same issue today "context deadline exceeded"
![image](https://user-images.githubusercontent.com/9293779/37139638-4ca1873a-22b7-11e8-8951-fdc4074351ff.png)

and also running
![image](https://user-images.githubusercontent.com/9293779/37139766-c3229f5c-22b7-11e8-8379-e493935fd3c0.png)

strange is the fact the dashboard seems to be active
![image](https://user-images.githubusercontent.com/9293779/37140141-5c266eda-22b9-11e8-8722-0bc1f2f921b0.png)

after a minikube delete; minikube start; pachctl deploy local; pachctl port-forward&
![image](https://user-images.githubusercontent.com/9293779/37142496-efda27be-22c0-11e8-885e-79ee9f8992a9.png)

it seems that all cmd line commands are working 
![image](https://user-images.githubusercontent.com/9293779/37142558-1e0dce42-22c1-11e8-92a2-c023441af759.png)


BUT the dashboard is dead (chrome and safari)
![image](https://user-images.githubusercontent.com/9293779/37142175-fd07616e-22bf-11e8-9c42-db915749ac03.png)


Any thoughts?! thanks!

**Comment by radu-adrian-moldovan** [comment]
well, I had to take it again from scratch and clear browser cache to have it working, strange!

**Comment by dwhitena** [comment]
@radu-adrian-moldovan Thanks for reporting. It seems like some of what you are hitting is around port-forwarding and other things that you are hitting are related to the browser cache. After redeploying Pachyderm you should run port forward again and do a hard refresh in the browser for your dashboard. That should clear up any of those issues. Also, if you command line connection to Pachyderm is flaky, you can always manually point `pachctl` to the cluster by setting the ADDRESS variable:

```
export ADDRESS=<k8s master IP>:30650
```

**Comment by suneeta-mall** [comment]
How does the the patchctl client connects to server? have no issue hitting the web interface but the patchctl client can not talk to server .. 
```
➜  pachyderm $ kubectl get all                                                                                                                 
NAME                           READY     STATUS    RESTARTS   AGE
pod/dash-5d974d8668-92p79      2/2       Running   0          14m
pod/etcd-0                     1/1       Running   0          14m
pod/pachd-75bfdb94c8-9ktr5     1/1       Running   0          14m

NAME                    TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                                   AGE
service/dash            NodePort    100.69.174.28    <none>        8080:30080/TCP,8081:30081/TCP                             14m
service/etcd            NodePort    100.67.145.145   <none>        2379:30985/TCP                                            14m
service/etcd-headless   ClusterIP   None             <none>        2380/TCP                                                  14m
service/kubernetes      ClusterIP   100.64.0.1       <none>        443/TCP                                                   4d
service/pachd           NodePort    100.70.44.204    <none>        650:30650/TCP,651:30651/TCP,652:30652/TCP,999:30999/TCP   14m

NAME                      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/dash      1         1         1            1           14m
deployment.apps/neltapi   2         2         2            2           2d
deployment.apps/pachd     1         1         1            1           14m

NAME                                 DESIRED   CURRENT   READY     AGE
replicaset.apps/dash-5d974d8668      1         1         1         14m
replicaset.apps/neltapi-6b4b4977b8   0         0         0         2d
replicaset.apps/neltapi-77c5c47876   0         0         0         2d
replicaset.apps/neltapi-7c8fc78485   2         2         2         2d
replicaset.apps/neltapi-7db7d8dbf8   0         0         0         2d
replicaset.apps/neltapi-7ff8548cff   0         0         0         2d
replicaset.apps/neltapi-8455b8f9bb   0         0         0         2d
replicaset.apps/pachd-75bfdb94c8     1         1         1         14m

NAME                    DESIRED   CURRENT   AGE
statefulset.apps/etcd   1         1         14m
➜  pachyderm$ pachctl version --verbose                                                                                                       
COMPONENT           VERSION             
pachctl             1.7.3               
context deadline exceeded
```
and I have already deployed it twice. 

I would think portforward is only convenience for us to access on local and the patchctl client would still use server master for communication.

**Comment by suneeta-mall** [comment]
I finally figured this error out! 

@JoeyZwicker Yes I agree .. this error is least explained. The documentation of having to specify the ADDRESS can be improved. I can raise a PR to spell that better if that makes sense but I am not sure why this value need to be set in the first place. Is not this already discoverable by kubeconfig/server host?  
The choice of port-forward and variable seems to fiddly to me .. IMO, pachctl should not have to use port-forward or ADDRESS but it should be discovered by config ... is there is a reason for why this choice was made?

**Comment by gabrielgrant** [comment]
@suneeta-mall i certainly agree this could be more user-friendly. I believe the issue is that, by default, a k8s cluster may have no route in to the pachd pod from the outside world (unless manually exposed), so port-forward is needed for us to piggy-back on the hole that k8s has by default for it's own API server

If/when a user has manually opened a route to their pachd pod (or, in the case of local deploy, where it is available by default), they need some way of telling pach where it should be trying to connect. This is where `ADDRESS` comes in. I believe this can also be specified in the pachyderm config file (`$HOME/.pachyderm/config.json`), though I'm not sure exactly how off the top of my head

I totally agree that this is pretty subtle/tricky to understand (esp. for newer k8s users). If you wanted to submit a PR with improved docs about this that would have helped you avoid getting stuck (or, better yet, suggest a better way of connecting that would require less explanation) I'd really appreciate it!

**Comment by suneeta-mall** [comment]
@gabrielgrant My apologies that I was not clear earlier. 
I was just indicating that instead of asking user to set the address variable pachctl should discover it itself using the current context. The ip of api server is discoverable @ ` dig +short api.<cluster-DNS>` and cluster DNS via kube context so it should be simpler and one less thing user need to do. 
Agreed on port-forward need, but it should be users choice.

**Comment by gabrielgrant** [comment]
@suneeta-mall fyi: submitted a feature request related to your suggestion about ADDRESS autodiscovery here: #3089
